File: server/apps/nexotype/schemas/omics/variant_schemas.py
from pydantic import BaseModel, ConfigDict
from datetime import datetime

# ==========================================
# Request Schemas
# ==========================================

class VariantCreate(BaseModel):
    """Schema for creating a new variant"""
    gene_id: int
    db_snp_id: str
    hgvs_c: str | None = None
    hgvs_p: str | None = None


class VariantUpdate(BaseModel):
    """Schema for updating a variant"""
    gene_id: int = None
    db_snp_id: str = None
    hgvs_c: str | None = None
    hgvs_p: str | None = None


# ==========================================
# Response Schemas
# ==========================================

class VariantDetail(BaseModel):
    """Schema for variant details"""
    id: int
    gene_id: int
    db_snp_id: str
    hgvs_c: str | None = None
    hgvs_p: str | None = None
    created_at: datetime
    updated_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)


class VariantListResponse(BaseModel):
    """Response schema for listing variants"""
    success: bool
    data: list[VariantDetail] | None = None
    count: int = 0
    error: str | None = None


class VariantResponse(BaseModel):
    """Response schema for single variant operations"""
    success: bool
    data: VariantDetail | None = None
    error: str | None = None


class MessageResponse(BaseModel):
    """Simple message response"""
    success: bool
    message: str | None = None
    error: str | None = None

File: client/src/modules/nexotype/schemas/omics/variant.schemas.ts
import { z } from 'zod';

// ==========================================
// Full Schema (GET responses)
// ==========================================

export const VariantSchema = z.object({
  id: z.number(),
  gene_id: z.number().int(),
  db_snp_id: z.string().min(1).max(20),
  hgvs_c: z.string().max(50).nullable(),
  hgvs_p: z.string().max(50).nullable(),
  created_at: z.string(),
  updated_at: z.string().nullable(),
});

// ==========================================
// Input Schemas (POST/PUT)
// ==========================================

export const CreateVariantSchema = z.object({
  gene_id: z.number().int(),
  db_snp_id: z.string().min(1).max(20),
  hgvs_c: z.string().max(50).nullable().optional(),
  hgvs_p: z.string().max(50).nullable().optional(),
});

export const UpdateVariantSchema = z.object({
  gene_id: z.number().int().optional(),
  db_snp_id: z.string().min(1).max(20).optional(),
  hgvs_c: z.string().max(50).nullable().optional(),
  hgvs_p: z.string().max(50).nullable().optional(),
});

// ==========================================
// Types
// ==========================================

export type Variant = z.infer<typeof VariantSchema>;
export type CreateVariant = z.infer<typeof CreateVariantSchema>;
export type UpdateVariant = z.infer<typeof UpdateVariantSchema>;

// ==========================================
// Response Types
// ==========================================

export type VariantResponse = {
  success: boolean;
  data?: Variant;
  error?: string;
};

export type VariantsResponse = {
  success: boolean;
  data?: Variant[];
  count?: number;
  error?: string;
};
